src/events/mqtt.js

```diff
@@ -97,10 +116,57 @@ function subscribeWrapper(originalSubscribeFunc) {
             tracer.addException(err);
         }
 
-        return originalSubscribeFunc.apply(this, [topic, options, patchedCallback]);
+        return originalSubscribeFunc.apply(this, [topic, internalOptions, patchedCallback]);
     };
 }
 
+/**
+ * Wraps the on command function with tracing
+ * @param {Function} originalOnFunc The wrapped function
+ * from mqtt module
+ * @returns {Function} The wrapped function
+ */
+function onWrapper(originalOnFunc) {
+    return function internalOnWrapper(funcName, callback) {
+        let patchedCallback = callback;
+        try {
+            if (funcName === 'message') {
+                const responseMetadata = {
+                    region: this.options.region,
+                    protocol: this.options.protocol,
+                    host: this.options.host,
+                };
+                const payload = {
+                    clientId: this.options.clientId,
+                    protocolId: this.options.protocolId,
+                    protocolVersion: this.options.protocolVersion,
+                };
+                const promise = new Promise((resolve) => {
+                    patchedCallback = (topic, message, ...rest) => {
+                        const { slsEvent: mqttEvent, startTime } = eventInterface.initializeEvent('mqtt', topic, 'onMessage', 'mqtt');
+                        payload.message = message ? message.toString() : message;
+                        eventInterface.finalizeEvent(
+                            mqttEvent,
+                            startTime,
+                            null,
+                            responseMetadata,
+                            payload
+                        );
+                        if (callback) {
+                            callback(topic, message, ...rest);
+                        }
+                        tracer.addEvent(mqttEvent, promise);
```

Comment:
no need for a promise here (you are not waiting for anything) - just remove the promise creation, resolve, and dont pass a promise parameter

Reply:
provide null or undefined nstead?

Reply:
undefined is good (just don't pass a second arg)